packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts

```diff
@@ -127,6 +128,21 @@ export function isHostElement(node: unknown): node is HTMLElement {
     return !isUndefined(InternalSlot.get(node));
 }
 
+// Return true if any descendant is a host element
+export function containsHost(node: Node) {
+    const descendants = arrayFromCollection(node.childNodes);
+    let descendant;
+    while (!isUndefined((descendant = descendants.shift()))) {
+        if (isHostElement(descendant)) {
+            return true;
+        }
+        // TypeScript does not like passing an array-like in as an array
+        // @ts-ignore
+        ArrayPush.apply(descendants, descendant.childNodes);
+    }
+    return false;
+}
```

Comment:
I'm not really happy with this solution. It's probably terrible for performance to keep traversing all children like this whenever `get innerHTML` is called. (If you're wondering why I didn't at least use a TreeWalker, it's because it's buggy in IE11.)

The reason we need this is because a light-DOM component could have deeply-nested elements with a synthetic shadow element inside. If that's the case, we can't call the native `get innerHTML` on any ancestor of that synthetic shadow element.

@jodarove We may be able to just enable the patches for all elements and avoid this check... What do you think about defaulting `ENABLE_ELEMENT_PATCH` to true?

Reply:
> If you're wondering why I didn't at least use a TreeWalker, it's because it's buggy in IE11.

Maybe worth adding this as a comment, it was my first reaction reading this code. I wasn't aware that `TreeWalker` was broken on IE11.

Reply:
After some difficulty, I managed to get TreeWalker working in IE.

Reply:
I'm still not sure about this code. The implementation will definitely be slow for deep DOM trees, since a call to `get innerHTML` will traverse the whole DOM, then call `innerHTML` for all children, traversing the DOM again for those children, then repeat...

I see two options:

1. We can say "You should never need to call `get innerHTML`/`get outerHTML`/etc in most cases. If it's expensive, you shouldn't use it."
2. We could universally patch `innerHTML`/`outerHTML`/etc without the expensive check, so it would apply even to DOM trees having nothing to do with LWC. (`ENABLE_ELEMENT_PATCH`)

Reply:
Actually it looks like I'm wrong; this doesn't recursively call the TreeWalker. It only calls it once, because the rest of the `innerHTML`/`outerHTML` calls go through the Polymer polyfills.

I benchmarked it (https://github.com/salesforce/lwc/commit/410cccf0c49f69698d677b1b0345127ca2d413fa) and even at a DOM depth of 1000 it only takes ~4ms on my machine, most of which is spent in `getOuterHTML`, which I think is acceptable.

![Screen Shot 2021-09-23 at 9 54 27 AM](https://user-images.githubusercontent.com/283842/134551095-5ba9f3d6-c5e7-4e2b-8ef1-0989ff0ac79f.png)